### PR TITLE
Output digests of build dependencies

### PR DESCRIPTION
**Pull request by shhsu**
**Purpose of the PR:**
Output digests of build dependencies.

**Fixes #** 
#25

@Azure/azure-container-registry

### CODE REVIEW

**Comment by northtyphoon** [comment]
@ehotinger

File: pkg/commands/docker.go [hunk 1]
```diff
@@ -125,3 +123,33 @@ func (t *dockerBuildTask) Push(runner build.Runner) error {
...
+			}
+		}
+	}
+	return nil
+}
+
+func queryDigest(runner build.Runner, reference *build.ImageReference) error {
+	refString := reference.String()
+	line, err := runner.QueryCmd("docker", []string{
+		"image", "ls", "--digests", "--format", "\"{{ .Digest }}\"", refString,
```
Comment (northtyphoon) [inline pkg/commands/docker.go]: {{ .Digest }} what is the extra space for?

File: pkg/grok/docker_compose.go [hunk 2]
```diff
@@ -113,15 +113,19 @@ func ResolveDockerComposeDependencies(env *build.BuilderContext, projectDirector
 		} else {
 			dockerfilePath = filepath.Join(imageContext, env.Expand(service.Build.Dockerfile))
 		}
-		runtime, buildtime, err := ResolveDockerfileDependencies(dockerfilePath)
+		runtime, buildtimes, err := ResolveDockerfileDependencies(dockerfilePath)
```
Comment (northtyphoon) [inline pkg/grok/docker_compose.go]: Nit: ImageDependencies use Buildtime, better be consistent

File: pkg/shell/runner.go [hunk 3]
```diff
@@ -65,14 +74,30 @@ func (r *shellRunner) executeCmd(obfuscate func([]string), cmdExe string, cmdArg
...
 }
 
-func (r *shellRunner) execute(cmd *exec.Cmd) error {
+func (r *shellRunner) execute(cmd *exec.Cmd, output io.Writer) error {
 	cmd.Env = append(cmd.Env, os.Environ()...)
 	cmd.Env = append(cmd.Env, r.context.Export()...)
 	cmd.Stdin = os.Stdin
-	cmd.Stdout = os.Stdout
+	if output != nil {
+		cmd.Stdout = io.MultiWriter(os.Stdout, output)
```
Comment (northtyphoon) [inline pkg/shell/runner.go]: Will it also write image digest query cmd output to os.stdout and eventually go to user's build log? It seems not necessary.

File: pkg/shell/runner.go [hunk 4]
```diff
@@ -65,14 +74,30 @@ func (r *shellRunner) executeCmd(obfuscate func([]string), cmdExe string, cmdArg
 		obfuscate(displayValues)
 	}
 	logrus.Infof("Running command %s %s", cmdExe, strings.Join(displayValues, " "))
-	return r.execute(cmd)
+	var output io.Writer
```
Comment (northtyphoon) [inline pkg/shell/runner.go]: I think if you change readOuputs from bool to io.Writer, and QueryCmd pass *bytes.Buffer and ExecuteCmd pass nil, the code will be cleaner.

File: pkg/commands/docker.go (hunk 1 above)
Comment (shhsu) [reply pkg/commands/docker.go]: I always leave the space when I do the query but I suppose I do not need it.

File: pkg/grok/docker_compose.go (hunk 2 above)
Comment (shhsu) [reply pkg/grok/docker_compose.go]: I usually name it buildtimes when I need to iterate. I guess I am not doing it here

File: pkg/shell/runner.go (hunk 4 above)
Comment (shhsu) [reply pkg/shell/runner.go]: sure

File: pkg/shell/runner.go (hunk 3 above)
Comment (shhsu) [reply pkg/shell/runner.go]: Not sure about this one. I always show command and its outputs and I think we should still show the outputs even if the output is captured by the builder and used.

File: pkg/build.go [hunk 5]
```diff
@@ -51,11 +53,69 @@ type Target interface {
...
+// ImageReference defines the reference to a docker image
+type ImageReference struct {
+	Registry   string `json:"registry"`
+	Repository string `json:"repository"`
+	Tag        string `json:"tag,omitempty"`
+	Digest     string `json:"digest"`
+	reference  reference.Reference
+}
+
+// NewImageReference parse a path of a image and create a ImageReference object
```
Comment (ehotinger) [inline pkg/build.go]: NIT: some wording throughout docs. (parse**s**, create**s**). Same comment applies for `String()`

**Comment by ehotinger** [comment]
LGTM. :shipit:

File: pkg/grok/docker_compose.go (hunk 2 above)
Comment (shhsu) [reply pkg/grok/docker_compose.go]: Fixed

**Pull request merged**